### ISSUE TITLE
Set implementation in JRuby 9.2 breaks interoperability with non-Set objects

### ISSUE THREAD

**Issue by ivoanjo** [jruby/jruby#5227]
### Environment

* JRuby: `jruby 9.2.0.0 (2.5.0) 2018-05-24 81156a8 OpenJDK 64-Bit Server VM 25.71-b00 on 1.8.0-internal-jenkins_2017_07_04_16_04-b00 [linux-x86_64]`
* Kernel: `Linux maruhime 4.13.0-45-generic #50-Ubuntu SMP Wed May 30 08:23:18 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
* Distro: `Ubuntu 17.10`

### Expected Behavior

The original Ruby `Set` implementation can, with a bit of ingenuity be "coerced" into interoperating with other objects.

Testcase:

```ruby
require 'set'

puts RUBY_DESCRIPTION

class FakeSet
  def is_a?(klass)
    super || klass == ::Set
  end

  def size
    0
  end

  def all?
    false
  end

  def any?
    false
  end
end

real_set = Set.new
fake_set = FakeSet.new

puts real_set < fake_set
puts real_set <= fake_set
puts real_set >= fake_set
puts real_set > fake_set
puts real_set.disjoint?(fake_set)
puts real_set.intersect?(fake_set)
```

Result:

```
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]
false
true
false
false
true
false
```

I rely on this for the [`Persistent💎::Set`](https://gitlab.com/ivoanjo/persistent-dmnd/blob/master/lib/persistent_dmnd/set.rb) implementation in my [Persistent-💎 ](https://gitlab.com/ivoanjo/persistent-dmnd) gem.

This works because the Ruby set implementation actually asks the _other_ object if it `is_a?(Set)`, thus allowing the object to "lie" about its identity, see <https://github.com/ruby/ruby/blob/v2_6_0_preview2/lib/set.rb#L258-L322>.

### Actual Behavior

With the new `Set` implementation on JRuby 9.2, this is the result of the testcase:

```ruby
jruby 9.2.0.0 (2.5.0) 2018-05-24 81156a8 OpenJDK 64-Bit Server VM 25.71-b00 on 1.8.0-internal-jenkins_2017_07_04_16_04-b00 [linux-x86_64]
ArgumentError: value must be a set
  proper_subset? at org/jruby/ext/set/RubySet.java:541
          <main> at testcase.rb:26
```

This happens because the `<`, `<=`, `>=`, `>`, `disjoint?` and `intersect?` Java-based implementations use `set instanceof RubySet` which unfortunately is not as object-oriented in Java as in Ruby, and thus the other object never gets a say on this matter, see <https://github.com/jruby/jruby/blob/9.2.0.0/core/src/main/java/org/jruby/ext/set/RubySet.java#L486-L581> and hence it breaks this corner case :)

The fix would probably be to first go for the `instanceof RubySet` check, but then call `is_a?` and allow the other object to have a say on the matter.

As usual, thanks for all the fish, and have a great weekend! :tada:

**Comment by ivoanjo** [comment]
Also `==` is on the same boat.

**Comment by headius** [comment]
So I guess the fix would be allowing non-native set objects to be used via dynamic dispatch, yes?

**Comment by headius** [comment]
I suppose another option would be leaving these methods in Ruby? Seems like we'd want the more efficient impl if possible, if these methods are used heavily.

**Comment by ivoanjo** [comment]
Yeah I don't think it's worth the performance impact of going back to Ruby code just to support this corner case (I am sorry for writing it too, in my defense the core Ruby code could be a lot more duck-typed in these cases, and it did work on every ruby implementation/version I tried it on lol), probably your first dynamic dispatch option would be the nicer way to go.

**Comment by headius** [comment]
I've pushed a possible fix that introduces a SetLike module into Set and delegates to those methods when the incoming set is not a RubySet. See https://github.com/jruby/jruby/tree/setlike.

```
 $ jruby blah.rb
jruby 9.2.1.0-SNAPSHOT (2.5.0) 2018-06-25 459b812 Java HotSpot(TM) 64-Bit Server VM 25.161-b12 on 1.8.0_161-b12 +jit [darwin-x86_64]
false
true
false
false
true
false
```

**Comment by headius** [comment]
Ping @kares

**Comment by kares** [comment]
this is just stupid - understandable but 'stupid'. at some point MRI will rewrite set.rb into native and might loose these semantics, guess time will tell whether Set#@hash details stop leaking and those methods might start only working for a set arg. doubt much real-world apps would break/use these, its a single library atm?

@headius no objections from me - it just seems like we might end up going back and forth, we'll see ...

**Comment by headius** [comment]
@kares I agree completely. This may even be worth bringing up with ruby-core, but I am betting they'd have no interest in locking down duck-typed scenarios like this. Generally I'd say overriding `is_a?` and friends are a bad idea, but I know that a ton of code does it.

For now, though, I think we need the fallback somehow. Do we have any concerns about mixing a special module in that doesn't exist in MRI's Set? The alternative would be to have these be "static" module methods and just dyncall them from our Java impl. It just means a bit more hopping around to reach the module, since we can't simply invoke against a superclass.

@ivoanjo I guess there's no specs for this? Writing some might be a good first place to start a discussion about whether this behavior is expected/specified/desirable.

**Comment by kares** [comment]
> Do we have any concerns about mixing a special module in that doesn't exist in MRI's Set?

fine by me - would maybe hide the module with `private_constant`.
alternatively you could just take those set.rb bits and put the into `class Set` - not sure but should work.
... since we do still "leak" `@hash` from the native Set impl for compatibility

**Comment by ivoanjo** [comment]
> since we do still "leak" @hash from the native Set impl for compatibility

Note that my gem does not rely on the `@hash` (not sure if that's why you were commenting on it, but I wanted to doublecheck 😄) -- it just presents a `Set` that is duck-type compatible with the regular `Set`, and the current `set.rb` MRI implementation just works because after doing the `is_a?(Set)` it then only uses the public interface to communicate with the "alien" Set. (Which, aside from how the type check is made, is similar to the current behavior of the JRuby set [where it has a fast path for unmodified `Set` and a "normal" one for the other cases](https://github.com/jruby/jruby/blob/9.2.0.0/core/src/main/java/org/jruby/ext/set/RubySet.java#L490-L497)).

It seems to me that if Ruby had a `.to_set` convention, it would be a drop-in replacement in this case; for instance by doing a simple `foo.respond_to?(:to_set) && set = foo.to_set`.

> @ivoanjo I guess there's no specs for this? Writing some might be a good first place to start a discussion about whether this behavior is expected/specified/desirable.

Yes totally @headius! I'm in for trying to fix/improve this.  As for writing the specs do you mean for the current behavior (e.g. document the current behavior the Persistent-💎 gem is relying on from MRI in Rubyspec) or something else?

**Comment by ivoanjo** [comment]
@headius thanks for the ping, I've had this on my to-do list for far too long ;)

I've just submitted the specs for the MRI behavior, see https://github.com/ruby/spec/pull/629

> You may also be interested in adding the missing behavior :-)

I have this also on my to-do list, but I'm not sure I'll be able to pick this up soon-ish 😭

**Comment by ivoanjo** [comment]
Issue raised upstream @ <https://bugs.ruby-lang.org/issues/15240>.
